### ISSUE TITLE
The new consumer can't consume msg from new partition when dynamic extend parition

### ISSUE THREAD

**Issue by DavidLiuXh** [apache/pulsar#6610]
**Describe the bug**
The new consumer can't consume msg from new partition when dynamic extend parition

**To Reproduce**
      1.  The pulsar cluster has three brokers;
      2.  Created partitioned topic whith three partition;
      3.  Run three consumers;
      4.  At this time, can see that each consumer consumes a parition;
      5.  Use 'pulsar-admin update-partitioned-topic' to dynamic extend partitions to four；
      6.  I can see that three consumer consumes four paritions;
      7.  Add a new consumer，but the new consumer can't consume msg from new 
          partition;  

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Centos 7
**Additional context**
- Pulasr 2.5.0
- bookkeeper 4.10.0

**Comment by sijie** [comment]
What is the subscription type are you using?

**Comment by DavidLiuXh** [comment]
> What is the subscription type are you using?

@sijie failover subscribe

**Comment by sijie** [comment]
@codelipenghui @jiazhai Can any of you help with this issue?

**Comment by codelipenghui** [comment]
@DavidLiuXh I try to reproduce the problem follow the steps that you provided, but I can't reproduce it. I have added a unit test #6699 for consumer distribution when partitions update. Please help check if it is consistent with what you described. Thanks.

**Comment by 315157973** [comment]
I also checked this problem before

@ DavidLiuXh's description is not accurate. This should be:
1 Create a Topic with 3 Partitions, and create 3 Consumers
2 Update the number of Partitions to 4
3 Add a new consumer
4 We expect that each consumer consumes one Partition, the reality is that 3 consumers consume 4 Partitions. Occasional discovery

the reason:
In failover mode, the sorting of consumers is based on the name, and the sorting result of the name is not fixed, it should be sorted by consumer id instead.

@codelipenghui

**Comment by codelipenghui** [comment]
@315157973 Oh, thanks. I will try to fix in #6699

**Comment by sijie** [comment]
@315157973 Are you using the same consumer name? If you don't provide consumer name, the consumer name is generated by broker, so the case you described shouldn't happen.

**Comment by 315157973** [comment]
@sijie  The automatically generated names are numbers + letters, and the ones generated later may be ranked first. You can reproduce it with a unit test

**Comment by 315157973** [comment]
Suppose there are 3 consumers, and their automatically generated names are: b1, b2, b4.

The consumer lists of the 3 nodes are：

node1： b1、b2、b4
node2： b1、b2、b4
node3： b1、b2、b4


org.apache.pulsar.broker.service.AbstractDispatcherSingleActiveConsumer#pickAndScheduleActiveConsumer

According to line 122

index = partitionIndex % consumersSize;

0 % 3 = 0
1 % 3 = 1
2 % 3 = 2

Partition0 is consumed by b1
Partition1 is consumed by b2
Partition2 is consumed by b4

When increased to 4 nodes, 4 consumers。The fourth consumer's name is b3

The consumer lists of the 4 nodes are：
node1： b1、b2、b3、b4
node2： b1、b2、b3、b4
node3： b1、b2、b3、b4
node4： b3

The consumption relationship becomes as follows：
0 % 4 = 0
1 % 4 = 1
2 % 4 = 2
3 % 1 = 0

Partition0 is consumed by b1
Partition1 is consumed by b2
Partition2 is consumed by b3
Partition3 is consumed by b3

b4 is out of work

Only after the client knows the Partition changes, b4 can enter the work。

**Comment by codelipenghui** [comment]
So, after the client knows new partition added, b4 eventually takes over the new partition, right?  Are you using java client?

**Comment by 315157973** [comment]
Both are right

**Comment by tisonkun** [comment]
Closed as stale. Please create a new issue if it's still relevant to the maintained versions.
